### PR TITLE
Add ACE Arsenal stats

### LINKED ISSUE #5964: ACE Arsenal stats suggestions
I am currently gathering informations in order to make the stats feature in ACE Arsenal,

### General

- Weight `mass` (bar + number)

### Primary / handgun:

- Rate of fire `reloadTime` (bar)
- Accuracy `dispersion` (bar)
- Range `maxZeroing` (bar)
- Impact `hit` (bar)

### Launchers

- Range `maxZeroing` (bar)
- Impact `hit` (bar)

### Containers

- Ballistic protection `passthrough` (bar)
- Explosion resistance `armor` (bar)
- Load `maximumLoad` (bar)

### Helmets

Same as containers minus load.

Any suggestions?

### PR DESCRIPTION
**Pull request by alganthe**
**When merged this pull request will:**
- Add ACE Arsenal stats

Stats documentation as well as a documentation overhaul for ACE Arsenal is coming in an other PR.

### Current stats:: 
<details>
  <summary>Click to expand</summary>

## Default stats
All tabs:
- Mass, kg / lbs (mass)

Left panel "misc" tabs + the right panel tabs only have mass by default

### Primary / handgun weapons
- Rate of fire, RPM  + bar (reloadTime)
- Accuracy, MIL / MOA + bar (dispersion)
- Range, bar (maxZeroing)
- Impact, bar (hit, initspeed)

### Launchers
- Range, bar (maxZeroing)
- Impact, bar (hit, initspeed)

### Uniform / vests
- Ballistic protection, bar (passthrough)
- Explosive resistance, bar (armor)
- Load, bar (maximumLoad)

### Backpacks
- Load, bar (maximumLoad)

### Headgear
- Ballistic protection, bar (passthrough)
- Explosive resistance, bar (armor)

## Ballistic stats
These stats will only show up if the ballistics module is loaded.

### Primary / handgun
- Barrel twist, mm / inches (ACE_barrelTwist)
- Barrel length, mm / inches (ACE_barrelLength)

### Mags
- Ammo, classname (ammo)
- Ballistic coefficient, ballistic coef / drag model (ACE_dragModel, ACE_ballisticCoefficients)
- Bullet mass, g / gr (ACE_bulletMass)

## Hearing stats
These stats will only show up if the hearing module is loaded.

### Headgear
- Hearing protection, bar (ACE_hearing_protection)
- Volume muffling, bar, (ACE_hearing_lowerVolume)

## Explosive stat
These stats will only show up if the explosives module is loaded.

### Right tab misc
- Explosive range, m / ft (ACE_explosive_range)

</details>

Close #5964

### CODE REVIEW

File: addons/arsenal/XEH_preInit.sqf [hunk 1]
```diff
@@ -1,13 +1,309 @@
...
+            }]],
+            [["ACE_barrelLength"], localize LSTRING(statBarrelLength), [false, true], [], [{}, {
+                params ["_stat", "_config"];
+
+                private _barrelLength = getNumber (_config >> _stat select 0);
+
+                format ["%1mm (%2in)",_barrelLength, (_barrelLength / 25.4) toFixed 1];
+            }]]
+        ];
+    } foreach [0, 1];
```
Comment (Dystopian) [inline addons/arsenal/XEH_preInit.sqf]: `foreach` -> `forEach` here and everywhere

File: addons/arsenal/ui/RscAttributes.hpp [hunk 2]
```diff
@@ -268,6 +268,195 @@ class GVAR(display) {
...
+                    idc = IDC_statsText5;
+                    y = QUOTE(50 * GRID_H);
+                };
+            };
+        };
+        class statsButton: ctrlButton {
+            idc = IDC_statsButton;
+            style= 2;
+            text=">";
+            onButtonClick = QUOTE(ARR_2([QQGVAR(statsButton), [ctrlparent (_this select 0)]]) call CBA_fnc_localEvent);
```
Comment (Dystopian) [inline addons/arsenal/ui/RscAttributes.hpp]: `ctrlparent` -> `ctrlParent` here and below

File: addons/arsenal/functions/fnc_handleStats.sqf [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+if !(isNil "_itemCfg") then {
+
+    private _handleStatsFnc = {
+        params ["_statsIndex", "_leftPanel"];
+        private ["_statsArray", "_statsList", "_isLeftPanel", "_currentPage"];
+
+        // Get the proper list and page
+        if (_leftPanel) then {
+            _isLeftPanel = true;
```
Comment (dedmen) [inline addons/arsenal/functions/fnc_handleStats.sqf]: you could return an array from the if statement here. And then use `params` on that array. Get's rid of that `private`

Comment (alganthe) [reply addons/arsenal/functions/fnc_handleStats.sqf]: totally true, plus I already do it somewhere else, I think it's for the loadouts list caching

**Comment by SilentSpike** [comment]
Can you put an image of the stats UI in the comments here please? 😁

**Comment by alganthe** [comment]
stats image

**Comment by SilentSpike** [comment]
Feedback on the UI: The only thing I don't really like is the red cross button. I think it would look much nicer if it fit with the rest of the UI and was black with a white cross (and square, with no margin around it so that it fit the corner of the container).

The editor UI has a button like this:

**Comment by SilentSpike** [comment]
Ideally it also shouldn't influence the centring of the page title text

**Comment by alganthe** [comment]
^ better?
img1
img2

**Comment by SilentSpike** [comment]
Yeah, much better 👍😄

Personally prefer the non-red version, but I think both are fine if others prefer that one

**Comment by alganthe** [comment]
it's not a "red version" it's me hovering hover it and showing up my title color :D

**Comment by alganthe** [comment]
Possible issue raised by dedmen:
- Other mods that append to the stats arrays at preInit may run before the arrays are created.
- If they run at postInit they won't be present in 3DEN (postInit doesn't run there)
- If I create the array at preStart it's the same issue for other mods appending / pushBack at preStart. Also would force to check if the stat was added if they add it in preInit.

**Comment by tbeswick96** [comment]
Use a CBA event?

**Comment by SilentSpike** [comment]
There should be an API function to add/remove stats which can handle these considerations internally

**Comment by alganthe** [comment]
True, i'll get on it, also variant 3 for the page indicator won :D

**Comment by SilentSpike** [comment]
The only other option I could see would be to add a config structure for stats, that way they're always present when an addon is loaded

**Comment by dedmen** [comment]
config is also much easier. No need for other mods to check "Does ACE arsenal exist? Does their addStat function exist?"

**Comment by alganthe** [comment]
So, run a func on preStart that builds the arrays from all the config entries and make an API to add / remove stats later? (preInit mostly due to 3DEN)

**Comment by alganthe** [comment]
And, the add / remove API is also done.

File: addons/ballistics/stringtable.xml [hunk 4]
```diff
@@ -2279,5 +2279,24 @@
...
             <Chinesesimp>[ACE] 弹药补给箱</Chinesesimp>
         </Key>
+        <Key ID="STR_ACE_Ballistics_statBarrelTwist">
+            <English>Barrel twist</English>
+        </Key>
+        <Key ID="STR_ACE_Ballistics_statBarrelLength">
+            <English>Barrel length</English>
+            <French>Longueur du canon</French>
+        </Key>
+        <Key ID="STR_ACE_Ballistics_statAmmo">
```
Comment (Dystopian) [inline addons/ballistics/stringtable.xml]: `str_dn_ammo` ?

File: addons/arsenal/CfgACEArsenalStats.hpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
+class CfgACEArsenalStats {
+    class statBase {
+        scope = 1;
+        priority = 0;
+        stats[]= {};
```
Comment (Dystopian) [inline addons/arsenal/CfgACEArsenalStats.hpp]: spaces around `=` and everywhere

File: addons/ballistics/stringtable.xml (hunk 4 above)
Comment (alganthe) [reply addons/ballistics/stringtable.xml]: ?

File: addons/arsenal/CfgACEArsenalStats.hpp (hunk 5 above)
Comment (alganthe) [reply addons/arsenal/CfgACEArsenalStats.hpp]: doesn't matter

Comment (jonpas) [reply addons/arsenal/CfgACEArsenalStats.hpp]: Yes it does.

File: addons/explosives/config.cpp [hunk 6]
```diff
@@ -45,3 +45,17 @@ class CfgMineTriggers {
         mineTriggerRange = 1;
     };
 };
+
+class CfgACEArsenalStats {
+    class statBase;
+    class explosiveRange: statBase {
+        scope = 2;
+        priority = 1;
+        stats[]= {"ace_explosives_Range"};
```
Comment (Dystopian) [inline addons/explosives/config.cpp]: QGVAR(Range) here and below other QGVARs

File: addons/ballistics/stringtable.xml (hunk 4 above)
Comment (Dystopian) [reply addons/ballistics/stringtable.xml]: `str_dn_ammo` `<English>Ammo</English>` - does it suit here?

File: addons/flashlights/config.cpp [hunk 7]
```diff
@@ -17,3 +17,17 @@ class CfgPatches {
 #include "CfgEventHandlers.hpp"
 #include "CfgVehicles.hpp"
 #include "CfgWeapons.hpp"
+
+class CfgACEArsenalStats {
```
Comment (Dystopian) [inline addons/flashlights/config.cpp]: should go to `CfgACEArsenalStats.hpp` this and others

File: addons/ballistics/stringtable.xml (hunk 4 above)
Comment (dedmen) [reply addons/ballistics/stringtable.xml]: I guess the already existing vanilla translation.

**Comment by shukari** [comment]
Is it possible to have in any progessbar the raw value. I see my guys with a lineal an compare lenghtes... 😃

**Comment by alganthe** [comment]
Added script profiler related macro and code, should be good to go now.

File: addons/arsenal/script_component.hpp [hunk 8]
```diff
@@ -5,6 +5,7 @@
 // #define DEBUG_MODE_FULL
 // #define DISABLE_COMPILE_CACHE
 // #define ENABLE_PERFORMANCE_COUNTERS
+// #define ENABLE_PERF_PROFILING
```
Comment (PabstMirror) [inline addons/arsenal/script_component.hpp]: maybe add a little note about what this does / requires

Comment (alganthe) [reply addons/arsenal/script_component.hpp]: true.

File: addons/arsenal/CfgACEArsenalStats.hpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
+class CfgACEArsenalStats {
```
Comment (jonpas) [inline addons/arsenal/CfgACEArsenalStats.hpp]: Should be renamed to `class ACE_Arsenal_Stats` (`GVAR(Stats)`) to fit other configs we have (`ACE_detector`, `ACE_UI`, `ACE_Settings`, `ACE_Extensions`, etc.).

File: addons/arsenal/CfgACEArsenalStats.hpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    class statBase {
+        scope = 1;
+        priority = 0;
+        stats[]= {};
+        displayName= "";
```
Comment (jonpas) [inline addons/arsenal/CfgACEArsenalStats.hpp]: Spacing around `=`, rest of this file and other config files too.

File: addons/arsenal/CfgACEArsenalStats.hpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        showBar= 0;
+        showText= 0;
+        barStatement = "";
+        textStatement = "";
+        condition = "true";
+        tabs[]= {{}, {}};
+    };
+    class bananaPotassium: statBase {
```
Comment (jonpas) [inline addons/arsenal/CfgACEArsenalStats.hpp]: We should prefix those too, this config will be public API.

I think it's best to prefix all, but some base ones that have nothing to do with ACE features could be left unprefixed.

File: addons/ballistics/functions/fnc_statTextStatement_weaponMuzzleVelocity.sqf [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+if (_magazine isEqualTo "") then {
+    _initSpeed
+} else {
+    private _ammoCfg = (configFile >> "CfgAmmo" >> (getText (configFile >> "CfgMagazines" >> _magazine >> "ammo")));
+    private _barrelLength = getNumber (_config >> "ACE_barrelLength");
+    private _muzzleVelocityTable = getArray (_ammoCfg >> "ACE_muzzleVelocities");
+    private _barrelLengthTable = getArray (_ammoCfg >> "ACE_barrelLengths");
+
+    if (_barrelLength != 0 && {count _muzzleVelocityTable > 0} && {count _barrelLengthTable > 0}) then {
+            private _muzzleVelocity = [_barrelLength, _muzzleVelocityTable, _barrelLengthTable, 0] call EFUNC(advanced_ballistics,calculateBarrelLengthVelocityShift);
```
Comment (jonpas) [inline addons/ballistics/functions/fnc_statTextStatement_weaponMuzzleVelocity.sqf]: Indentation much?

File: addons/ballistics/functions/fnc_statTextStatement_weaponMuzzleVelocity.sqf [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    } else {
+        [primaryWeapon EGVAR(arsenal,center), ""]
+    };
+} else {
+    private _primaryMag = handgunMagazine EGVAR(arsenal,center);
+
+    if (count _primaryMag > 0) then {
+        [handgunWeapon EGVAR(arsenal,center), _primaryMag select 0]
+    } else {
+        [handgunWeapon EGVAR(arsenal,center), ""]
```
Comment (jonpas) [inline addons/ballistics/functions/fnc_statTextStatement_weaponMuzzleVelocity.sqf]: Couldn't just use `param`?

File: addons/arsenal/functions/fnc_onSelChangedRight.sqf [hunk 14]
```diff
@@ -14,6 +14,10 @@
 #include "script_component.hpp"
 #include "..\defines.hpp"
 
+#ifdef ENABLE_PERF_PROFILING
+    private _scopeOnSelChangedRight= createProfileScope QFUNC(onSelChangedRight);
```
Comment (jonpas) [inline addons/arsenal/functions/fnc_onSelChangedRight.sqf]: What's with your spaces around `=`.

File: addons/arsenal/functions/fnc_onSelChangedRightListnBox.sqf [hunk 15]
```diff
@@ -14,6 +14,10 @@
 #include "script_component.hpp"
 #include "..\defines.hpp"
 
+#ifdef ENABLE_PERF_PROFILING
+    private _scopeOnSelChangedRightLNB= createProfileScope QFUNC(onSelChangedRightListnBox);
```
Comment (jonpas) [inline addons/arsenal/functions/fnc_onSelChangedRightListnBox.sqf]: Spaces around `=`.

File: addons/arsenal/CfgACEArsenalStats.hpp (hunk 11 above)
Comment (alganthe) [reply addons/arsenal/CfgACEArsenalStats.hpp]: what kind of prefix tho.

Comment (jonpas) [reply addons/arsenal/CfgACEArsenalStats.hpp]: `ACE_`

**Comment by alganthe** [comment]
Todo: fix scaling on 4:3 for the stats box, it's fucked right now.

**Comment by alganthe** [comment]
yep.

**Pull request merged**